libcudacxx/docs/extended_api/asynchronous_operations/memcpy_async_tx.md

```diff
@@ -0,0 +1,100 @@
+---
+grand_parent: Extended API
+parent: Asynchronous Operations
+---
+
+# `cuda::device::memcpy_async_tx`
+
+Defined in header `<cuda/barrier>`:
+
+```cuda
+template <typename T, size_t Alignment>
+inline __device__
+async_contract_fulfillment
+cuda::device::memcpy_async_tx(
+  T* dest,
+  const T* src,
+  cuda::aligned_size_t<Alignment> size,
+  cuda::barrier<cuda::thread_scope_block>& bar);
+```
+
+Copies `size` bytes from global memory `src` to shared memory `dest` and decrements the transaction count of `bar` by `size` bytes.
+
+## Preconditions
+
+* `src`, `dest` are 16-byte aligned and `size` is a multiple of 16, i.e.,
+  `Alignment >= 16`.
+* `dest` points to a shared memory allocation that is at least `size` bytes wide.
+* `src` points to a global memory allocation that is at least `size` bytes wide.
+* `bar` is located in shared memory
+* If either `destination` or `source` is an invalid or null pointer, the
+    behavior is undefined (even if `count` is zero).
+
+## Requires
+
+* `is_trivially_copyable_v<T>` is true.
+ 
+## Notes
+
+This function can only be used under CUDA Compute Capability 9.0 (Hopper) or
+higher.
+
+There is no feature flag to check if `cuda::device::memcpy_async_tx` is
+available.
+
+**Comparison to `cuda::memcpy_async`**: `memcpy_async_tx` supports a subset of
+the operations of `memcpy_async`. It gives more control over the synchronization
+with a barrier than `memcpy_async`. Currently, `memcpy_async_tx` has no synchronous
+fallback mechanism., i.e., it currently does not work on older hardware
+(pre-CUDA Compute Capability 9.0, i.e., Hopper).
+
+## Return Value
+
+Returns `async_contract_fulfillment::async`.
```

Comment:
We should not guarantee that we always return `async`. 

Where do we document `async_contract_fulfillment` ?

Reply:
Why not?

> Where do we document async_contract_fulfillment ?

This is a bit weird. The memcpy_async functions return this type, but are documented as returning void. Perhaps we should do the same here?

Reply:
> Why not?

Because then lifting any of the restrictions in the current implementation becomes only possible if the `async` value continues to be returned.

> The memcpy_async functions return this type, but are documented as returning void. Perhaps we should do the same here?

The change to return this type was supposed to happen as part of the Hopper feature exposure. @wmaxey @griwes @miscco may know why some `memcpy_async` functions returns it but the docs don't capture that. 

My preference is to only document the return type here, and document what the values of this type mean in the documentation of the type (e.g. open a github issue to document `async_contract_fulfillment`).

Reply:
Given that we do not expect the return value to carry any information going further we went ahead and removed it from the documentation to be consistent with the other memcpy_async functions